adapter/internal/sourcewatcher/sourcewatcher.go

```diff
@@ -0,0 +1,365 @@
+func compareRepository(localRepository *git.Repository){
+	remote, err := localRepository.Remote("origin")
+	if err != nil{
+		loggers.LoggerSourceWatcher.ErrorC(logging.ErrorDetails{
+			Message: fmt.Sprintf("Error while returning remote : %s", err.Error()),
+			Severity: logging.CRITICAL,
+			ErrorCode: 2504,
+		})
+	}
+
+	gitAuth, err := auth.GetGitAuth()
+	if err != nil {
+		loggers.LoggerSourceWatcher.ErrorC(logging.ErrorDetails{
+			Message: fmt.Sprintf("Error while authenticating with the remote repository : %s", err.Error()),
+			Severity: logging.CRITICAL,
+			ErrorCode: 2502,
+		})
+	}
+
+	remoteList, err := remote.List(&git.ListOptions{
+		Auth: gitAuth,
+	})
+	if err != nil {
+		loggers.LoggerSourceWatcher.ErrorC(logging.ErrorDetails{
+			Message: fmt.Sprintf("Error while listing remote : %s", err.Error()),
+			Severity: logging.MAJOR,
+			ErrorCode: 2505,
+		})
+	}
+
+	ref, err := localRepository.Head()
+	if err != nil {
+		loggers.LoggerSourceWatcher.ErrorC(logging.ErrorDetails{
+			Message: fmt.Sprintf("Error while retrieving local repository head : %s", err.Error()),
+			Severity: logging.MAJOR,
+			ErrorCode: 2506,
+		})
+	}
+
+	refName := ref.Name()
+
+	localRepositoryHash := ref.Hash().String()
+
+	for _, r := range remoteList {
+		if r.Name() == refName && !r.Hash().IsZero() && localRepositoryHash != r.Hash().String() {
+			loggers.LoggerSourceWatcher.Info("Fetching commit with hash: ", r.Hash().String(), " from remote repository")
+			pullChanges(localRepository)
+
+			err := processArtifactChanges()
+			if err != nil {
+				loggers.LoggerSourceWatcher.ErrorC(logging.ErrorDetails{
+					Message: fmt.Sprintf("Error while processing artifact changes : %s", err.Error()),
+					Severity: logging.MAJOR,
+					ErrorCode: 2507,
+				})
+			}
+
+			//Redeploy changes
+			artifactsMap, err = api.ProcessMountedAPIProjects()
+			if err != nil {
+				loggers.LoggerSourceWatcher.ErrorC(logging.ErrorDetails{
+					Message: fmt.Sprintf("Processing API artifacts failed : %s", err.Error()),
+					Severity: logging.CRITICAL,
+					ErrorCode: 2501,
+				})
+				return
+			}
+		}
+	}
+
+}
+
+// pullChanges pulls changes from the given repository
+func pullChanges(localRepository *git.Repository){
+	conf, err := config.ReadConfigs()
+	if err != nil {
+		loggers.LoggerSourceWatcher.ErrorC(logging.ErrorDetails{
+			Message: fmt.Sprintf("Error reading configs : %s", err.Error()),
+			Severity: logging.BLOCKER,
+			ErrorCode: 2500,
+		})
+		return
+	}
+
+	branch := conf.Adapter.SourceControl.Repository.Branch
+
+	gitAuth, err := auth.GetGitAuth()
+	if err != nil {
+		loggers.LoggerSourceWatcher.ErrorC(logging.ErrorDetails{
+			Message: fmt.Sprintf("Error while authenticating with the remote repository : %s", err.Error()),
+			Severity: logging.CRITICAL,
+			ErrorCode: 2502,
+		})
+	}
+
+	pullOptions := &git.PullOptions{
+		Auth: gitAuth,
+	}
+
+	if branch != "" {
+		pullOptions.ReferenceName = plumbing.ReferenceName("refs/heads/" + branch)
```

Comment:
let's create a constant for `refs/heads/` as it is used in other places.

Reply:
Fixed with 5eadc80